### PR TITLE
Display warning on try-runtime spec_name mismatch

### PR DESCRIPTION
**Pull request by kianenigma**
Fixes https://github.com/paritytech/substrate/issues/9576

Note that we don't want to abort execution if there is a mismatch. For example, in substrate we want to only execute the migrations of one pallet over the state of polkadot. In this case, it is totally expected that the spec_names don't match. 

Example: 

```
2021-08-20 09:31:12.390  WARN main try_runtime_cli: version mismatch: remote spec name: kusama, expected (local chain spec, aka. `--chain`): development
```

### CODE REVIEW

File: utils/frame/remote-externalities/src/rpc_api.rs [hunk 1]
```diff
@@ -88,3 +88,24 @@ async fn build_client<S: AsRef<str>>(from: S) -> Result<WsClient, String> {
 		.await
 		.map_err(|e| format!("`WsClientBuilder` failed to build: {:?}", e))
 }
+
+/// Get the runtime of a given chain.
```
Comment (chevdor) [inline utils/frame/remote-externalities/src/rpc_api.rs:92]:
```suggestion
/// Get the runtime version of a given chain.
```

File: utils/frame/try-runtime/cli/src/lib.rs [hunk 2]
```diff
@@ -274,6 +279,8 @@ where
 		max_runtime_instances,
 	);
 
+	check_spec_name::<Block>(shared.url.clone(), config.chain_spec.name().to_string()).await;
```
Comment (chevdor) [inline utils/frame/try-runtime/cli/src/lib.rs:282]: We are already checking L204. Do we need to check again?

File: utils/frame/try-runtime/cli/src/lib.rs [hunk 3]
```diff
@@ -484,3 +493,33 @@ fn extract_code(spec: Box<dyn ChainSpec>) -> sc_cli::Result<(StorageKey, Storage
...
+
+/// Check the spec_name of an `ext`
+///
+/// If the version does not exist, or if it does not match with the given, it emits a warning.
+async fn check_spec_name<Block: BlockT + serde::de::DeserializeOwned>(
+	uri: String,
+	expected_spec_name: String,
+) {
+	let expected_spec_name = expected_spec_name.to_lowercase();
+	// let expected_runtime_str: sp_runtime::RuntimeString = expected_spec_name.into();
```
Comment (chevdor) [inline utils/frame/try-runtime/cli/src/lib.rs:505]:
```suggestion
```

File: utils/frame/try-runtime/cli/src/lib.rs [hunk 4]
```diff
@@ -484,3 +493,33 @@ fn extract_code(spec: Box<dyn ChainSpec>) -> sc_cli::Result<(StorageKey, Storage
...
+	match remote_externalities::rpc_api::get_runtime_version::<Block, _>(uri.clone(), None)
+		.await
+		.map(|version| <RuntimeString as Into<String>>::into(version.spec_name.clone()))
```
Comment (bkchr) [inline utils/frame/try-runtime/cli/src/lib.rs:508]:
```suggestion
		.map(|version| String::from(version.spec_name.clone()))
```

File: utils/frame/try-runtime/cli/src/lib.rs (hunk 3 above)
Comment (bkchr) [inline utils/frame/try-runtime/cli/src/lib.rs:505]:
```suggestion
```

**Review by bkchr — APPROVED** [review]

**Review by chevdor — APPROVED** [review]
LGTM besides the minor points I mentioned

File: utils/frame/try-runtime/cli/src/lib.rs (hunk 2 above)
Comment (kianenigma) [reply utils/frame/try-runtime/cli/src/lib.rs:282]: yes, they are different functions and different call paths.

**Comment by kianenigma** [comment]
bot merge

**Pull request merged**